### PR TITLE
Add --target-rid top level option

### PR DESCRIPTION
**Pull request by am11**
Closes https://github.com/dotnet/source-build/issues/4413

cc @tmds

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -12,6 +12,7 @@ usage()
   echo "Common settings:"
   echo "  --binaryLog                     Create MSBuild binary log (short: -bl)"
   echo "  --configuration <value>         Build configuration: 'Debug' or 'Release' (short: -c)"
+  echo "  -r, --target-rid <value>        Runtime identifier to be used by the build. e.g. alpine.3.18-arm64, fedora.37-x64, freebsd.13-arm64, ubuntu.19.10-x64"
```
Comment (am11) [inline build.sh:15]:
```suggestion
  echo "  --target-rid, -r <value>        Runtime identifier to be used by the build. e.g. alpine.3.18-arm64, fedora.37-x64, freebsd.13-arm64, ubuntu.19.10-x64"
```

File: build.sh [hunk 2]
```diff
@@ -280,6 +286,10 @@ fi
 source $scriptroot/eng/common/native/init-os-and-arch.sh
 source $scriptroot/eng/common/native/init-distro-rid.sh
 initDistroRidGlobal "$os" "$arch" ""
+if [[ -n "$target_rid" ]]; then
+  __DistroRid="$target_rid"
```
Comment (am11) [inline build.sh:290]:
```suggestion
  # overwrite with user's preferred RID
  export __DistroRid="$target_rid"
```

File: build.sh [hunk 3]
```diff
@@ -12,6 +12,7 @@ usage()
   echo "Common settings:"
   echo "  --binaryLog                     Create MSBuild binary log (short: -bl)"
   echo "  --configuration <value>         Build configuration: 'Debug' or 'Release' (short: -c)"
+  echo "  --target-rid, -r <value>        Runtime identifier to be used by the build. e.g. alpine.3.18-arm64, fedora.37-x64, freebsd.13-arm64, ubuntu.19.10-x64"
```
Comment (tmds) [inline build.sh]: Suggestion to make more clear what this rid does, and what it does not do (https://github.com/dotnet/source-build/issues/4419):

```suggestion
  echo "  --rid, -r <value>               Overrides the rid that is produced by the build. e.g. alpine.3.18-arm64, fedora.37-x64, freebsd.13-arm64, ubuntu.19.10-x64"
```

File: build.sh [hunk 4]
```diff
@@ -280,6 +286,11 @@ fi
 source $scriptroot/eng/common/native/init-os-and-arch.sh
 source $scriptroot/eng/common/native/init-distro-rid.sh
 initDistroRidGlobal "$os" "$arch" ""
+if [[ -n "$target_rid" ]]; then
+  # overwrite with user's preferred RID
```
Comment (am11) [inline build.sh:290]:
```suggestion
  # override with user's preferred RID
```

File: build.sh [hunk 5]
```diff
@@ -280,6 +286,11 @@ fi
 source $scriptroot/eng/common/native/init-os-and-arch.sh
 source $scriptroot/eng/common/native/init-distro-rid.sh
 initDistroRidGlobal "$os" "$arch" ""
+if [[ -n "$target_rid" ]]; then
+  # override with user's preferred RID
+  export __DistroRid="$target_rid"
```
Comment (tmds) [inline build.sh]:
```suggestion
```

I think we can leave this out.

File: build.sh [hunk 6]
```diff
@@ -100,6 +102,10 @@ while [[ $# > 0 ]]; do
       configuration=$2
       shift
       ;;
+    -target-rid|-r)
```
Comment (tmds) [inline build.sh:105]:
```suggestion
    -rid|-r)
```

Comment (am11) [reply build.sh:105]: I thought settled on --target-rid?

File: build.sh (hunk 5 above)
Comment (am11) [reply build.sh]: That affects OutputRID calculation in Directory.Build.props.

Comment (tmds) [reply build.sh]: I'm not sure what you mean. For runtime, we're passing the rid down here:

https://github.com/dotnet/runtime/blob/40787436cba1b732a2507aaba2cd2a1e6f809a50/eng/DotNetBuild.props#L58

If there are places where `__DistroRid` overrides `TargetRid` then we should fix them.

File: build.sh (hunk 6 above)
Comment (tmds) [reply build.sh:105]: I think _target_ adds to the confusion of https://github.com/dotnet/source-build/issues/4419.
For this top-level argument, imo we can just name it `--rid`.

File: build.sh (hunk 5 above)
Comment (am11) [reply build.sh]: I think you are right, once `OutputRid` is overriden, we only use `__DistroRid` for outputOS calculation in `Directory.Build.props`: https://grep.app/search?q=__DistroRid&filter[repo][0]=dotnet/runtime

Comment (am11) [inline build.sh:291]:
```suggestion
```

File: build.sh (hunk 6 above)
Comment (tmds) [reply build.sh:105]: @ViktorHofer thoughts?

**Comment by am11** [comment]
@tmds, I think we can get rid of calculation in this repo: https://github.com/dotnet/dotnet/blob/3999c972cc2e8330a922e40a60ca89545756d450/Directory.Build.props#L54 It is only used for runtime where we have this calculation. Besides it's the old logic borrowed from runtime.

File: build.sh (hunk 6 above)
Comment (ViktorHofer) [reply build.sh:105]: I would prefer target-rid here with a good documentation explaining that this really just determines the name and nothing else. I feel that consistency is more important here.

Btw should we add the switch to the Windows script as well?

And last but not least this PR needs to target the sdk repo.

Comment (tmds) [reply build.sh:105]: fwiw, I know _target_ is meant to be similar as how gcc has a _build_, _host_ and _target_ platform.
In contrast to that, a target rid does not describe a target platform. It is just a name.
I think that difference is what causes the confusion in https://github.com/dotnet/source-build/issues/4419.

Comment (am11) [reply build.sh:105]: > this PR needs to target the sdk repo.

@ViktorHofer how would that work? VMR build.sh is different than the one in SDK.

Comment (ViktorHofer) [reply build.sh:105]: https://github.com/dotnet/sdk/blob/main/src/SourceBuild/content/build.sh

We don't accept any PRs in this repository. This is just a mirror of code from various repos. Eventually the VMR will become its own source-of-truth but not until the beginning of .NET 10. See https://github.com/dotnet/dotnet?tab=readme-ov-file#contribution

Comment (am11) [reply build.sh:105]: Ah, didn't knew it. Will close it.

**Comment by tmds** [comment]
> @tmds, I think we can get rid of calculation in this repo: 

We need this in case the user doesn't specify a rid. This is responsible to make non-portable builds use the rid calculated by `init-distro-rid.sh`.

**Comment by am11** [comment]
Closing in favor of https://github.com/dotnet/sdk/pull/41540.

**Pull request closed without merging**